### DISCUSSION TITLE
possibility of optimizing CRNN using other library optimizers

### DISCUSSION THREAD

**Discussion by yewalenikhil65** [DENG-MIT/CRNN#6] [Ideas]
https://discourse.julialang.org/t/ann-fluxopttools/25888
This link talks about using other Optim library optimisers along with Flux.

**Comment by jiweiqi** [comment]
Thanks for posting the link. It is fairly straightforward to apply Optim.jl's optimizer here. I have done it before.
Of particular interest is the L-BFGS optimizer, while it is uncertain that when L-BFGS will perform better, here is the downside of BFGS, it is not so clear how to do with mini-batching and BFGS prefers a monthly loss function.

**Reply by yewalenikhil65** [reply]
I thought `update!` considers only optimisers from Flux. 
Was there any noticeable difference in optimization for reaching global minima as compared to ADAM/SGD?
It came to my mind because some examples from DiffEqFlux suggest using `BFGS` after `ADAM`. It seems good enough for doing simple regression from the link posted but i am not sure how it fares with Neural-nets in general

**Reply by jiweiqi** [reply]
you don't have to call `update!` You can work with `p` directly.

The optimization with BFGS is empirical. In later stages, using BFGS does make the loss decrease further, but in general, SGD is more stable. I personally would like to stay with SGD most of the time, especially, when the model is close to a deep neural network.

**Comment by jiweiqi** [comment]
I feel https://www.mathworks.com/help/deeplearning/ref/trainlm.html;jsessionid=ff2565ad89f67e8126aee1162866 might perform better than Adam for moderate size of network. There is also coupe of packages in Julia has implemented it but not in Optim.Ji.

If you are interested in it, we can talk about the details and have a try.

**Reply by yewalenikhil65** [reply]
Hi @jiweiqi , Sorry for huge delay. 
But now I can come back to CRNN and previous discussions .  Regarding Levenberg-Marquardt algorithm(LM), I too saw recently found  JuliaNLSolvers /LsqFit.jl which uses fitting using LM.
Lets try this first on a non-stiff model shall we ?

I have recently created a Jump(Gillespie) study of https://en.wikipedia.org/wiki/Smoluchowski_coagulation_equation which is a non-stiff case and expect CRNN to work easily on it. Take a look here at a short tutorial i created for sciML/Catalyst https://github.com/yewalenikhil65/Population-balance-equation. 
We can get "data" from there for now, though in reality/practical aggregation models have only measurements of 1st species (thats what I could surmise from experimental papers about aggregation)

**Reply by yewalenikhil65** [reply]
> Hi @jiweiqi , Sorry for huge delay.
> But now I can come back to CRNN and previous discussions . Regarding Levenberg-Marquardt algorithm(LM), I too saw recently found JuliaNLSolvers /LsqFit.jl which uses fitting using LM.
> Lets try this first on a non-stiff model shall we ?
> 
> I have recently created a Jump(Gillespie) study of https://en.wikipedia.org/wiki/Smoluchowski_coagulation_equation which is a non-stiff case and expect CRNN to work easily on it. Take a look here at a short tutorial i created for sciML/Catalyst https://github.com/yewalenikhil65/Population-balance-equation.
> We can get "data" from there for now, though in reality aggregation models have only measurements of 1st species in https://en.wikipedia.org/wiki/Smoluchowski_coagulation_equation

I was going to write you on this myself today anyway..  We could treat "data" from Jump/Gillespie as noisy data(as gillespie algo is a random process itself and the states/solutions are in integers already) and try to learn the weights..I am creating a repository now for this problem. Shall I add send you the github invite as contributor ?

**Reply by jiweiqi** [reply]
Hi @yewalenikhil65 ... I am curious if CRNN can represent the population balance equations since it is not chemical reactions in the conventional sense.

For the implementation of LM, we should definitely try some example that is very simple. We can try Smoluchowski_coagulation_equation.

I am not sure if the LsqFit has a good LM implementation. As mentioned in SciML https://diffeq.sciml.ai/v1.8/analysis/parameter_estimation.html, LM performs badly, but this could be due to the implementation of LM. There is a legacy package for LM called minpack. Here is a Julia wrapper https://github.com/sglyon/MINPACK.jl. I think this one might be a better choice.

A note for LM is that we cannot use mini-batching with LM. Instead, we should use full-batch. This is not a problem for a small system. In addition, we could use distributed computing to parallel the evaluation of Jacobian.

**Reply by jiweiqi** [reply]
For the GitHub repo, yeah, it will be helpful to grant me access.

**Reply by jiweiqi** [reply]
It is good to start from synthesized data assuming all species are available. For practical consideration, it is another story, as we have discussed before, one can inference the model uncertainty.

**Reply by jiweiqi** [reply]
I have never dealt with stochastic systems before and I am excited about it. I am also working on some biology systems those days with biologists.

**Reply by yewalenikhil65** [reply]
> Hi @yewalenikhil65 ... I am curious if CRNN can represent the population balance equations since it is not chemical reactions in the conventional sense.
> 
True, they are not represented as chemical reactions conventionally but an aggregation process. But there are many literatures that regard PBEs as chemical reactions following **law of mass action** .(Partly the reason to model them as gillespie processes).
Only thing is the reaction rate constants are not some fixed-constants but dependent on volumes of particles that collide in so-called "bi-molecular reaction".So its not of much importance maybe to learn reaction rate constants accurately.

> For the implementation of LM, we should definitely try some example that is very simple. We can try Smoluchowski_coagulation_equation.
> 
I will create a data-set for this and add to repo. The way you did in case1/2/3.

> I am not sure if the LsqFit has a good LM implementation. As mentioned in SciML https://diffeq.sciml.ai/v1.8/analysis/parameter_estimation.html, LM performs badly, but this could be due to the implementation of LM. There is a legacy package for LM called minpack. Here is a Julia wrapper https://github.com/sglyon/MINPACK.jl. I think this one might be a better choice.
> 
Yes, i will take a look.. This seems like `fsolve` command in MATLAB
> A note for LM is that we cannot use mini-batching with LM. Instead, we should use full-batch. This is not a problem for a small system. In addition, we could use distributed computing to parallel the evaluation of Jacobian.
>
Ahh yes, Jacobian of cost function needed in LM .That might make it expensive . But i am not much aware of parallel computing in any language.

**Reply by jiweiqi** [reply]
Distributed computing in Julia is pretty straightforward but put a decorator@distributed` at the beginning of for loop.

**Reply by yewalenikhil65** [reply]
> I have never dealt with stochastic systems before and I am excited about it. I am also working on some biology systems those days with biologists.

After you paper got published, it had occured to me how would
`X' = JumpProblem(CRNN(X),params)` perform !. It will be expensive as `ODEProblems` are generally not expensive. I will create an option of having data from stochastic as well as deterministic(and noiseless) for now,

**Reply by jiweiqi** [reply]
Indeed, this is very exciting.
For stochastic systems, there should be some approximated inference methods that don't solve the gradient accurately but are fast. Those approaches are good to try as well.

**Reply by yewalenikhil65** [reply]
> Indeed, this is very exciting.
> For stochastic systems, there should be some approximated inference methods that don't solve the gradient accurately but are fast. Those approaches are good to try as well.

Hey @jiweiqi , I have invited you as contributor to https://github.com/yewalenikhil65/PBE_CRNN . 
Currently its private, lets make it public after it works. `PBE_generateReactions.jl` generates reactions and its corresponding `ODESystem`, which is included in CRNN file. Its quite explanatory as i have adopted most of code from case 1 of CRNN.

Trouble with this idea of `X' = JumpProblem(CRNN(X),params)` is that JumpProblem accepts reactionSystems/Discrete Problems and not `DEProblem` I think. so its difficult to make `JumpProblem` accept user-defined function like `crnn!`. Therefore for now, I have used `ODEProblem` for this Population balance eqs.

**Reply by yewalenikhil65** [reply]
My initial attempts of CRNN for this problem.. using ADAM doesn't seem to progress much. But maybe needs more tweaking. Do take  a look when you feel free.

**Reply by jiweiqi** [reply]
Sure. If adam doesn't work, the second-order optimizer will probably not working either. I don't think the issue is in the optimizer.

**Reply by jiweiqi** [reply]
By the way, here is a sample code I used to test lmfit.
```Julia
using LsqFit
using ForwardDiff
using Plots
using Flux

xdata = 0.0:1.0:10.0
xdata = xdata'
ydata = Float32.(vec(sin.(xdata)))

nn = Chain(
        Dense(1, 5, tanh),
        Dense(5, 5, tanh),
        Dense(5, 1))
x, re = Flux.destructure(nn)

function f(p)
    ypred = vec(re(p)(xdata))
    return @. abs(ypred - ydata)
end
f(x)

g = function (p)
    return ForwardDiff.jacobian(x -> f(x), p)
end
g(x)

fit = LsqFit.lmfit(f, g, x, Float64[]; show_trace=true, maxIter=1000, x_tol=1e-6)

display(fit.param)

plt = scatter(xdata[:], ydata[:], label="data");
plot!(plt, xdata[:], vec(re(fit.param)(xdata)), label="pred");
png(plt, "lmnn")
```

**Reply by yewalenikhil65** [reply]
> By the way, here is a sample code I used to test lmfit.
> 


Thanks @jiweiqi  , I am taking a look.. By the way, i had a little query related to CRNN. After training and learning the reactions, do you think predictions take lesser time that actually solving the ODEs(assuming we know them as we have done for using synthesized data) . Basically i wanted to after successful training do you think
`@btime ODEProblem(crnn!, u ,p) <  @ btime ODEProblem(trueODEfunc, u, p)` ?

**Reply by jiweiqi** [reply]
For small systems, I don't think there should be any difference as both are very fast. (say the number of species is less than 100).
If the system is super large, that's possible as CRNN computes dudt in matrix form.
So I don't expect it in the system that currently looked into.

**Reply by yewalenikhil65** [reply]
> For small systems, I don't think there should be any difference as both are very fast. (say the number of species is less than 100).
> If the system is super large, that's possible as CRNN computes dudt in matrix form.
> So I don't expect it in the system that currently looked into.

Ok.. and about  https://github.com/yewalenikhil65/PBE_CRNN  not fitting.. now its fitting well.. I made a mistake in forcing the bounds `ub` earlier. Will update the repo again.. just waiting to see how weights are evolving..

update:
i think the loss won't minimize further now .. take a look at learned weights and loss function here 
https://github.com/yewalenikhil65/PBE_CRNN/tree/main/main

**Reply by jiweiqi** [reply]
* Looks like the system is stiff from the value of w_b ?
* Consider a smaller learning rate, like 0.001

There are other things that can accelerate the code, which I will try when I got time next week.

**Reply by yewalenikhil65** [reply]
>     * Looks like the system is stiff from the value of w_b ?
> 
`Tsit5()` seems working fine though. I am not sure about stiffness.. Could be moderately stiff ?
expected `w_b` as follows seems similar to case 1, so I thought it's non-stiff
```julia
  6×5 Array{Int64,2}:
 -2   1   0   0  0
 -1  -1   1   0  0
 -1   0  -1   1  0
  0  -2   0   1  0
 -1   0   0  -1  1
  0  -1  -1   0  1
```
>     * Consider a smaller learning rate, like 0.001
> 
Ok
> 
> There are other things that can accelerate the code, which I will try when I got time next week.
>
ok . no problem

**Reply by jiweiqi** [reply]
What's your expected rate constants?

**Reply by yewalenikhil65** [reply]
Ohh sorry. I mistook `w_b` for `w_out`
Reaction rate constants depend on cluster size `N` i wish to consider..large `N` means more species and reactions and hence the distribution of rate constants is also wide-spread..i.e. as we increase `N` rate constants differ gradually in orders.. not suddenly

for `N=5` , for which we tried, I Suppose this case amounts to non-stiff ?
`k =  3.0599999999999996e-7 ,4.59e-7, 6.119999999999999e-7,6.119999999999999e-7, 7.65e-7, 7.65e-7`

lets increase N
for N=10
`1×25 LinearAlgebra.Adjoint{Float64,Array{Float64,1}}:`
`k =  3.06e-7  4.59e-7  6.12e-7  6.12e-7  7.65e-7  7.65e-7  9.18e-7  …  1.377e-6  1.53e-6  1.53e-6  1.53e-6  1.53e-6  1.53e-6`

for N=100
`1×2500 LinearAlgebra.Adjoint{Float64,Array{Float64,1}}:`
 `k = 3.06e-7  4.59e-7  6.12e-7  6.12e-7  7.65e-7  …  1.53e-5  1.53e-5  1.53e-5  1.53e-5  1.53e-5`

**Reply by jiweiqi** [reply]
looks like the system is not very stiff, but with some stiffness.

**Reply by yewalenikhil65** [reply]
Oh.. Does this mean I will have to try something from `robertson` or other cases than case 1 for this problem ?

**Reply by jiweiqi** [reply]
For N = 5, the system is not stiff. (This can be seen as the k are in the same orders).
For the code, you have written, the learned CRNN fit the data very well. In this case, it is likely that the model is not sensitive to the data you have generated. Considering put all species appearing in initial conditions, instead of only have species 1 in initial conditions.

**Reply by yewalenikhil65** [reply]
>Considering put all species appearing in initial conditions, instead of only have species 1 in initial conditions.
>
I think partial initialisation among species (say for N=5, if we initialise 1,2, and/or 3) should be in accordance with realistic situations as higher polymers are usually not present. But i will try it anyway

Another thing.. we know for sure that qualitatively higher species also follow similar trend as that for lower N values , even after increasing `N`. Can this knowledge help for larger `N` when stiffness increases ?

**Reply by jiweiqi** [reply]
Yeah, those tips are for testing the algorithm. We shall increase difficulties step by step.

As for "realistic", for a nonlinear model, it is in general difficult to inference model parameters if we don't carefully design the experiments. So theoretically, we eventually learn realistic data but probably not in the current way of randomly sampling a couple of initial conditions.

So, just relax everything you can for now to make sure the algorithm and code work.

**Reply by jiweiqi** [reply]
Here is an example of LM optimization of a neural network using MINPACK
https://gist.github.com/jiweiqi/ecb8fd57728f333dea4cf56a2a28d877

**Comment by yewalenikhil65** [comment]
Bdw.. @jiweiqi 
Have you seen this https://joshday.github.io/SparseRegression.jl/stable/  ?

**Reply by jiweiqi** [reply]
Hi @yewalenikhil65  I have not seen this package. But I know sparse regression. I have also talked about the difference between CRNn and SINDY in the CRNN paper. The key message is that, if you can list all of the possible pathways and the number is not high, SINDY can fit in, otherwise, we shall try neural networks like CRNN. In addition, if the rate constants are not constant, say temperature-dependent, I haven't figured out how SINDY will work for that, as the model is nonlinear now.

**Reply by jiweiqi** [reply]
By the way, sparse regression not working with stiff systems.

**Comment by jiweiqi** [comment]
@yewalenikhil65 I found the code I wrote for Robertson's problem using LM optimizer, although the code not working well
```Julia
using DiffEqFlux, OrdinaryDiffEq, Flux, Optim, Plots
using DifferentialEquations
using DiffEqSensitivity
using Zygote
using ForwardDiff
using LinearAlgebra
using Random
using Statistics
using ProgressBars, Printf
using Flux.Optimise: update!
using Flux.Losses: mae
using BSON: @save, @load
using LatinHypercubeSampling
using LsqFit

is_restart = false;
n_epoch = 2000;
n_plot = 10;

opt = ADAMW(0.005, (0.9, 0.999), 1.f-6);
datasize = 40;
batchsize = 16;
n_exp_train = 20;
n_exp_val = 5;
n_exp = n_exp_train + n_exp_val; 
noise = 1.f-4;
ns = 3;
nr = 6;

grad_max = 10 ^ (0.5);
maxiters = 10000;

# alg = AutoTsit5(Rosenbrock23(autodiff=false));
alg = Rosenbrock23(autodiff=false);
atol = [1e-6, 1e-8, 1e-6];
rtol = [1e-3, 1e-3, 1e-3];
lb = 1e-8;
ub = 1.f1;

np = nr * (2 * ns + 1) + 1;
p = (rand(Float32, np) .- 0.5) * 2 * sqrt(6 / (ns + nr));
p[end] = 1.e-1;

# Generate datasets
u0_list = rand(Float32, (n_exp, ns)) .* 2 .+ 0.5;
u0_list[:, 2:2] .= 0 .+ lb;
u0_list[:, [1, 3]] .= randomLHC(n_exp, 2) ./ n_exp .+ 0.5

tsteps = 10 .^ range(0, 5, length=datasize);
tspan = Float32[0, tsteps[end]];
t_end = tsteps[end]

k = [4.f-2, 3.f7, 1.f4];
ode_data_list = zeros(Float32, (n_exp, ns, datasize));
yscale_list = [];

function trueODEfunc(dydt, y, k, t)
    r1 = k[1] * y[1]
    r2 = k[2] * y[2] * y[2]
    r3 = k[3] * y[2] * y[3]
    dydt[1] = -r1 + r3
    dydt[2] = r1 - r2 - r3
    dydt[3] = r2
end

u0 = u0_list[1, :];
prob_trueode = ODEProblem(trueODEfunc, u0, tspan, k);

function max_min(ode_data)
    return maximum(ode_data, dims=2) .- minimum(ode_data, dims=2)
end

for i = 1:n_exp
    u0 = u0_list[i, :]
    prob_trueode = ODEProblem(trueODEfunc, u0, tspan, k)
    ode_data = Array(solve(prob_trueode, alg, saveat=tsteps, atol=atol, rtol=rtol))
    ode_data += randn(size(ode_data)) .* ode_data .* noise
    ode_data_list[i, :, :] = ode_data
    push!(yscale_list, max_min(ode_data))
end

yscale = maximum(hcat(yscale_list...), dims=2);
dydt_scale = yscale[:, 1] ./ t_end
show(stdout, "text/plain", round.(yscale', digits=8))

function p2vec(p)
    slope = abs(p[end])
    w_b = @view(p[1:nr]) .* (10 * slope)

    w_in = reshape(@view(p[nr * (ns + 1) + 1:nr * (2 * ns + 1)]), ns, nr)

    w_out = reshape(@view(p[nr + 1:nr * (ns + 1)]), ns, nr)
    w_out = @. -w_in * (10 ^ w_out)

    w_in = clamp.(w_in, 0, 2.5)
    return w_in, w_b, w_out
end

function display_p(p)
    w_in, w_b, w_out = p2vec(p)
    println("species (column) reaction (row)")
    println("w_in | w_b | w_out")
    display(hcat(w_in', w_b, w_out'))

    println("w_out_scale")
    w_out_ = (w_out .* dydt_scale)' .* exp.(w_b)
    # display(w_out_)
    # display(maximum(abs.(w_out_), dims=2)')
    display(w_out_ ./ maximum(abs.(w_out_), dims=2))
    println("slope = $(p[end])")
end
display_p(p)

function crnn(du, u, p, t)
    w_in_x = w_in' * @. log(clamp(u, lb, Inf))
    du .= w_out * (@. exp(w_in_x + w_b)) .* dydt_scale
end

u0 = u0_list[1, :]
prob = ODEProblem(crnn, u0, tspan, saveat=tsteps, atol=atol, rtol=rtol)

sense = BacksolveAdjoint(checkpointing=true; autojacvec=false);
function predict_neuralode(u0, p; sample = datasize)
    global w_in, w_b, w_out = p2vec(p)
    _prob = remake(prob, tspan=[0, tsteps[sample]])
    sol = solve(prob, alg, u0=u0, p=p, saveat=tsteps[1:sample], 
                sensalg=sense, verbose=false, maxiters=maxiters)
    pred = Array(sol)

    if sol.retcode == :Success
        nothing
    else
        println("ode solver failed")
    end
    return pred
end
pred = predict_neuralode(u0, p);

function loss_neuralode(p, i_exp; sample = datasize)
    pred = predict_neuralode(u0_list[i_exp, :], p; sample)
    ode_data = ode_data_list[i_exp, :, 1:size(pred)[2]]
    loss = mae(ode_data ./ yscale, pred ./ yscale)
    return loss
end
loss_neuralode(p, 1)

cbi = function (p, i_exp)
    ode_data = ode_data_list[i_exp, :, :]
    pred = predict_neuralode(u0_list[i_exp, :], p)
    l_plt = []
    for i = 1:ns
        plt = scatter(tsteps, ode_data[i, :], xscale=:log10,
                      markercolor=:transparent, label=string("data"))
        plot!(plt, tsteps[1:size(pred)[2]], pred[i, :], xscale=:log10, label=string("pred"))
        ylabel!(plt, "y$i")
        if i == ns
            xlabel!(plt, "Time [s]")
            plot!(plt, legend=:topleft)
        else
            plot!(plt, legend=false)
        end
        push!(l_plt, plt)
    end
    plt_all = plot(l_plt..., framestyle=:box, layouts = (ns, 1))
    png(plt_all, string("figs/i_exp_", i_exp))

    return false
end

l_loss_train = []
l_loss_val = []
l_grad = []
iter = 1
cb = function (p, loss_train, loss_val, g_norm)
    global l_loss_train, l_loss_val, l_grad, iter
    push!(l_loss_train, loss_train)
    push!(l_loss_val, loss_val)
    push!(l_grad, g_norm)

    if iter % n_plot == 0
        display_p(p)

        l_exp = randperm(n_exp)[1:1]
        println("update plot for ", l_exp)
        for i_exp in l_exp
            cbi(p, i_exp)
        end

        plt_loss = plot(l_loss_train, xscale=:identity, yscale=:log10, label="train")
        plot!(plt_loss, l_loss_val, xscale=:identity, yscale=:log10, label="val")
        plt_grad = plot(l_grad, xscale=:identity, yscale=:log10, label="grad_norm")
        xlabel!(plt_loss, "Epoch")
        xlabel!(plt_grad, "Epoch")
        ylabel!(plt_loss, "Loss")
        ylabel!(plt_grad, "Grad Norm")
        ylims!(plt_loss, (-Inf, 1))
        plt_all = plot([plt_loss, plt_grad]..., legend=:top)
        png(plt_all, "figs/loss_grad")

        @save "./checkpoint/mymodel.bson" p opt l_loss_train l_loss_val l_grad iter
    end
    iter += 1
end

if is_restart
    @load "./checkpoint/mymodel.bson" p opt l_loss_train l_loss_val l_grad iter
    iter += 1
    # opt = ADAMW(0.001, (0.9, 0.999), 1.f-6)
end

function loss_lm(p)
    [loss_neuralode(p, i) for i in 1:n_exp_train]
end
loss_lm(p);

g = function (p)
    return ForwardDiff.jacobian(x -> loss_lm(x), p)
end
g(p);

epochs = ProgressBar(iter:n_epoch);
loss_epoch = zeros(Float32, n_exp);
grad_norm = zeros(Float32, n_exp_train);
for epoch in epochs
    global p
    for i_exp in randperm(n_exp_train)
        sample = rand(batchsize:datasize)
        grad = ForwardDiff.gradient(x -> loss_neuralode(x, i_exp; sample), p)
        grad_norm[i_exp] = norm(grad, 2)
        if grad_norm[i_exp] > grad_max
            grad = grad ./ grad_norm[i_exp] .* grad_max
        end
        update!(opt, p, grad)
    end
    for i_exp in 1:n_exp
        loss_epoch[i_exp] = loss_neuralode(p, i_exp)
    end
    loss_train = mean(loss_epoch[1:n_exp_train]);
    loss_val = mean(loss_epoch[n_exp_train + 1:end]);
    g_norm = mean(grad_norm)
    set_description(epochs, string(@sprintf("Loss train %.4e val %.4e gnorm %.4e", loss_train, loss_val, g_norm)))
    cb(p, loss_train, loss_val, g_norm);

    if loss_train < 0.1
        break
    end
end

fit = LsqFit.lmfit(loss_lm, g, p, Float64[]; show_trace=true, maxIter=2000, x_tol=1e-8)
p_fit = fit.param;
display_p(p_fit)
cbi(p_fit, 1)
cbi(p_fit, n_exp)

# @printf("min loss train %.4e val %.4e\n", minimum(l_loss_train), minimum(l_loss_val))

# for i_exp in 1:n_exp
#     cbi(p, i_exp)
# end

```

**Reply by yewalenikhil65** [reply]
This seems to do LsqFit **after** we had done our fitting using `ADAM` ?

**Reply by jiweiqi** [reply]
Yes, I have seen many people use adam first and then second order optimizer. It is just heuristics.

**Comment by jiweiqi** [comment]
I tried LM for case 2 at https://gist.github.com/jiweiqi/4a485b90c5f706eeacd1e7f150a9688d. This time, I use the package of MINPACK.
For case 2 with Arrhenius law and thus could be stiff and highly nonlinear, the lm performs really bad comparing to Adam. The lm is also unstable comparing to Adam.
Maybe something could be further tuned to make it work.

**Reply by yewalenikhil65** [reply]
I tried applying same code of LM for case 1. but am facing error at `fsolve!` line

```julia
ArgumentError: Must have at least as many variables as equations
lmder(::typeof(f!), ::typeof(g!), ::Array{Float64,1}, ::Int64, ::Float64, ::Bool, ::Bool, ::Int64; _n::Int64, gtol::Float64, ftol::Float64, xtol::Float64, epsfcn::Float64, mode::Int64, nprint::Int64, maxfev::Int64, factor::Float64, ldfjac::Int64) at wrappers.jl:459
lmder(::Function, ::Function, ::Array{Float64,1}, ::Int64, ::Float64, ::Bool, ::Bool, ::Int64) at wrappers.jl:455
#fsolve#2 at MINPACK.jl:147 [inlined]
(::MINPACK.var"#fsolve##kw")(::NamedTuple{(:iterations, :tol, :show_trace, :tracing, :method),Tuple{Int64,Float64,Bool,Bool,Symbol}}, ::typeof(fsolve), ::Function, ::Function, ::Array{Float64,1}, ::Int64) at MINPACK.jl:144
top-level scope at case1.jl:229
```

**Reply by jiweiqi** [reply]
It is a limitation of the package which requires the number of fvec is greater than the number of parameters.

By the way, personally, I will not invest in lm algorithm those days since it looks indeed not working.

**Comment by yewalenikhil65** [comment]
@jiweiqi Well, since second order methods dont seem to give us push, i would say `mixed integer programming` could be worth a try ? https://github.com/JuliaOpt/Pajarito.jl or other `JuMP/MathOptINterface` solvers

**Reply by jiweiqi** [reply]
It depends on the scale. For a system with a handful of reactions and species, those optimizations will work as done many previous work.

For large system they won't

**Reply by yewalenikhil65** [reply]
Ohk,i would take your word for it.. meanwhile, you would be interested to take a look at this https://arxiv.org/abs/1902.04920 .Its based on loss function as `maximizing the likelihood function of the trajectory data`.. if you see page 21 bottom and page 22., you can see that they have been successful with stiff reaction system as well.

**Reply by jiweiqi** [reply]
Thanks for sharing.
1) do they use a predefined reaction library, like in SINDY? If that, I am not surprised that it works since you already know the reactions.
2) where did they say it is a stiff system? (since they don't use gradient descent, their algorithm should not suffer from stiffness)

**Reply by jiweiqi** [reply]
They are using Bayesian inference for learning models, for sure they should use likelihood.

**Reply by yewalenikhil65** [reply]
> Thanks for sharing.
> 
>     1. do they use a predefined reaction library, like in SINDY? If that, I am not surprised that it works since you already know the reactions.
> 
yes, they add number of possible `ratelaws` as basis.. like number of polynomial terms..and rely on sparse-regression like SINDy. I didnt understand how it is different from SINDy also.
>     2. where did they say it is a stiff system? (since they don't use gradient descent, their algorithm should not suffer from stiffness)
> 
table 13 says.. rate constants of their example was k = [0.25  0.001   0.3   100.0   2.0   0.1]
so i thought its possibly stiff
>They are using Bayesian inference for learning models, for sure they should use likelihood.
>
Yes. they use logarithmic likelihood. i didnt understand it quite though
